Runtime error for select on partitioned table with Global Index / IndexFullScan

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```mysql
drop table if exists t;
create table t (a int, b varchar(255), c varchar (255), primary key (a,b), unique key (c) global) partition by list columns (a,b) (partition p0 values in ((1,"1"),(2,"2"),(3,"3")), partition p1 values in ((100,"100"),(101,"101"),(102,"102"),DEFAULT));
insert into t values (1,1,1),(2,2,2),(101,101,101),(102,102,102);
select * from t;
explain select * from t;
```

### 2. What did you expect to see? (Required)
succeeded select

### 3. What did you see instead (Required)
```
tidb> select * from t;
ERROR 1105 (HY000): runtime error: index out of range [2] with length 2
```

Full output, including plan:
```mysql
tidb> create table t (a int, b varchar(255), c varchar (255), primary key (a,b), unique key (c) global) partition by list columns (a,b) (partition p0 values in ((1,"1"),(2,"2"),(3,"3")), partition p1 values in ((100,"100"),(101,"101"),(102,"102"),DEFAULT));
Query OK, 0 rows affected (0.00 sec)

tidb> insert into t values (1,1,1),(2,2,2),(101,101,101),(102,102,102);
Query OK, 4 rows affected (0.00 sec)
Records: 4  Duplicates: 0  Warnings: 0

tidb> select * from t;
ERROR 1105 (HY000): runtime error: index out of range [2] with length 2
tidb> explain select * from t;
+-----------------------+----------+-----------+---------------------+--------------------------------+
| id                    | estRows  | task      | access object       | operator info                  |
+-----------------------+----------+-----------+---------------------+--------------------------------+
| IndexReader_7         | 10000.00 | root      | partition:all       | index:IndexFullScan_6          |
| └─IndexFullScan_6 | 10000.00 | cop[tikv] | table:t, index:c(c) | keep order:false, stats:pseudo |
+-----------------------+----------+-----------+---------------------+--------------------------------+
2 rows in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
```
tidb_version(): Release Version: v8.5.0-alpha-31-g241c10294b
Edition: Community
Git Commit Hash: 241c10294b9c0305db590576700c00de306be573
Git Branch: master
UTC Build Time: 2024-11-03 21:45:52
GoVersion: go1.23.2
Race Enabled: false
Check Table Before Drop: false
Store: unistore
```

Full stack:
```
[2024/11/03 22:53:31.168 +01:00] [ERROR] [coprocessor.go:1158] ["copIteratorWork meet panic"] [r="runtime error: index out of range [2] with length 2"] ["stack trace"="github.com/pingcap/tidb/pkg/store/copr.(*copIteratorWorker).handleTask.func1
	/Users/devuser/repos/tidb/pkg/store/copr/coprocessor.go:1160
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:785
runtime.goPanicIndex
	/usr/local/go/src/runtime/panic.go:115
github.com/pingcap/tidb/pkg/tablecodec.decodeRestoredValuesV5
	/Users/devuser/repos/tidb/pkg/tablecodec/tablecodec.go:868
github.com/pingcap/tidb/pkg/tablecodec.reEncodeHandleConsiderNewCollation
	/Users/devuser/repos/tidb/pkg/tablecodec/tablecodec.go:835
github.com/pingcap/tidb/pkg/tablecodec.decodeIndexKvForClusteredIndexVersion1
	/Users/devuser/repos/tidb/pkg/tablecodec/tablecodec.go:1865
github.com/pingcap/tidb/pkg/tablecodec.DecodeIndexKV
	/Users/devuser/repos/tidb/pkg/tablecodec/tablecodec.go:990
github.com/pingcap/tidb/pkg/store/mockstore/unistore/cophandler.(*indexScanExec).Process
	/Users/devuser/repos/tidb/pkg/store/mockstore/unistore/cophandler/mpp_exec.go:296
github.com/pingcap/tidb/pkg/store/mockstore/unistore/tikv/dbreader.(*DBReader).Scan
	/Users/devuser/repos/tidb/pkg/store/mockstore/unistore/tikv/dbreader/db_reader.go:240
github.com/pingcap/tidb/pkg/store/mockstore/unistore/cophandler.(*indexScanExec).open
	/Users/devuser/repos/tidb/pkg/store/mockstore/unistore/cophandler/mpp_exec.go:349
github.com/pingcap/tidb/pkg/store/mockstore/unistore/cophandler.mppExecute
	/Users/devuser/repos/tidb/pkg/store/mockstore/unistore/cophandler/cop_handler.go:229
github.com/pingcap/tidb/pkg/store/mockstore/unistore/cophandler.buildAndRunMPPExecutor
	/Users/devuser/repos/tidb/pkg/store/mockstore/unistore/cophandler/cop_handler.go:220
github.com/pingcap/tidb/pkg/store/mockstore/unistore/cophandler.handleCopDAGRequest
	/Users/devuser/repos/tidb/pkg/store/mockstore/unistore/cophandler/cop_handler.go:175
github.com/pingcap/tidb/pkg/store/mockstore/unistore/cophandler.HandleCopRequestWithMPPCtx
	/Users/devuser/repos/tidb/pkg/store/mockstore/unistore/cophandler/cop_handler.go:102
github.com/pingcap/tidb/pkg/store/mockstore/unistore/cophandler.HandleCopRequest
	/Users/devuser/repos/tidb/pkg/store/mockstore/unistore/cophandler/cop_handler.go:90
github.com/pingcap/tidb/pkg/store/mockstore/unistore/tikv.(*Server).Coprocessor
	/Users/devuser/repos/tidb/pkg/store/mockstore/unistore/tikv/server.go:625
github.com/pingcap/tidb/pkg/store/mockstore/unistore.(*RPCClient).SendRequest
	/Users/devuser/repos/tidb/pkg/store/mockstore/unistore/rpc.go:261
github.com/pingcap/tidb/pkg/store/mockstore.(*clientRedirector).SendRequest
	/Users/devuser/repos/tidb/pkg/store/mockstore/redirector.go:72
github.com/tikv/client-go/v2/tikv.(*CodecClient).SendRequest
	/Users/devuser/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.8-0.20241023023120-691e80ae0ea9/tikv/test_util.go:62
github.com/tikv/client-go/v2/internal/client.interceptedClient.SendRequest.func1
	/Users/devuser/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.8-0.20241023023120-691e80ae0ea9/internal/client/client_interceptor.go:57
github.com/tikv/client-go/v2/internal/client.buildResourceControlInterceptor.func1.1
	/Users/devuser/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.8-0.20241023023120-691e80ae0ea9/internal/client/client_interceptor.go:124
github.com/tikv/client-go/v2/internal/client.interceptedClient.SendRequest
	/Users/devuser/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.8-0.20241023023120-691e80ae0ea9/internal/client/client_interceptor.go:58
github.com/tikv/client-go/v2/internal/client.reqCollapse.SendRequest
	/Users/devuser/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.8-0.20241023023120-691e80ae0ea9/internal/client/client_collapse.go:74
github.com/tikv/client-go/v2/internal/locate.(*RegionRequestSender).sendReqToRegion
	/Users/devuser/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.8-0.20241023023120-691e80ae0ea9/internal/locate/region_request.go:1114
github.com/tikv/client-go/v2/internal/locate.(*RegionRequestSender).SendReqCtx
	/Users/devuser/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.8-0.20241023023120-691e80ae0ea9/internal/locate/region_request.go:870
github.com/tikv/client-go/v2/txnkv/txnsnapshot.(*ClientHelper).SendReqCtx
	/Users/devuser/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.8-0.20241023023120-691e80ae0ea9/txnkv/txnsnapshot/client_helper.go:146
github.com/pingcap/tidb/pkg/store/copr.(*copIteratorWorker).handleTaskOnce
	/Users/devuser/repos/tidb/pkg/store/copr/coprocessor.go:1281
github.com/pingcap/tidb/pkg/store/copr.(*copIteratorWorker).handleTask
	/Users/devuser/repos/tidb/pkg/store/copr/coprocessor.go:1171
github.com/pingcap/tidb/pkg/store/copr.(*copIteratorWorker).run
	/Users/devuser/repos/tidb/pkg/store/copr/coprocessor.go:825"]
```

--- a/pkg/tablecodec/tablecodec.go
+++ b/pkg/tablecodec/tablecodec.go
@@ -832,13 +832,13 @@ func reEncodeHandleConsiderNewCollation(handle kv.Handle, columns []rowcodec.Col
 	if len(restoreData) == 0 {
 		return cHandleBytes, nil
 	}
-	// Remove global index extra column,
-	// because the type of `model.ExtraPhysTblID` always is int
-	// which means docode restore is not needed.
-	if len(columns) > 0 && columns[len(columns)-1].ID == model.ExtraPhysTblID {
-		columns = columns[:len(columns)-1]
+	// Remove some extra columns(ID < 0), such like `model.ExtraPhysTblID`.
+	// They are not belong to common handle and no need to restore data.
+	idx := len(columns)
+	for idx > 0 && columns[idx-1].ID < 0 {
+		idx--
 	}
-	return decodeRestoredValuesV5(columns, cHandleBytes, restoreData)
+	return decodeRestoredValuesV5(columns[:idx], cHandleBytes, restoreData)
 }
 
 func decodeRestoredValues(columns []rowcodec.ColInfo, restoredVal []byte) ([][]byte, error) {

--- a/pkg/tablecodec/tablecodec.go
+++ b/pkg/tablecodec/tablecodec.go
@@ -832,6 +832,12 @@ func reEncodeHandleConsiderNewCollation(handle kv.Handle, columns []rowcodec.Col
 	if len(restoreData) == 0 {
 		return cHandleBytes, nil
 	}
+	// Remove global index extra column,
+	// because the type of `model.ExtraPhysTblID` always is int
+	// which means docode restore is not needed.
+	if len(columns) > 0 && columns[len(columns)-1].ID == model.ExtraPhysTblID {
+		columns = columns[:len(columns)-1]
+	}
 	return decodeRestoredValuesV5(columns, cHandleBytes, restoreData)
 }
 

--- a/tests/integrationtest/r/globalindex/misc.result
+++ b/tests/integrationtest/r/globalindex/misc.result
@@ -238,3 +238,12 @@ t	CREATE TABLE `t` (
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
 PARTITION BY KEY (`b`) PARTITIONS 3
 drop table t;
+drop table if exists t;
+create table t (a int, b varchar(255), c varchar (255), primary key (a,b), unique key (c) global) partition by list columns (a,b) (partition p0 values in ((1,"1"),(2,"2"),(3,"3")), partition p1 values in ((100,"100"),(101,"101"),(102,"102"),DEFAULT));
+insert into t values (1,1,1),(2,2,2),(101,101,101),(102,102,102);
+select * from t;
+a	b	c
+1	1	1
+101	101	101
+102	102	102
+2	2	2

--- a/tests/integrationtest/t/globalindex/misc.test
+++ b/tests/integrationtest/t/globalindex/misc.test
@@ -152,3 +152,11 @@ show create table t;
 alter table t partition by key (b) partitions 3;
 show create table t;
 drop table t;
+
+# TestRestoreValuedWithGlobalIndex
+drop table if exists t;
+create table t (a int, b varchar(255), c varchar (255), primary key (a,b), unique key (c) global) partition by list columns (a,b) (partition p0 values in ((1,"1"),(2,"2"),(3,"3")), partition p1 values in ((100,"100"),(101,"101"),(102,"102"),DEFAULT));
+insert into t values (1,1,1),(2,2,2),(101,101,101),(102,102,102);
+--sorted_result
+select * from t;
+

